### DISCUSSION TITLE
Alternative ways to render components in lit-html

### DISCUSSION THREAD

**Discussion by alarbada** [lit/lit#3882] [Q&A]
So, currently the lit typescript plugin doesn't work for the latest typescript version. And, to be honest, without it the development experience is not great:

```typescript
    render() {
        // No typechecking, no goto definition working
        return html`<my-component something="a"> </my-component>`
    }
```

I was considering an alternative to get better typescript support, without needing external tooling. This is the main idea that I got, a custom directive:

```typescript
import {TemplateResult, nothing} from 'lit'
import {html, unsafeStatic} from 'lit/static-html.js';

function el(tag: string, props: Record<string, string | number>, children?: TemplateResult<1>) {
	let attrs = ''
	for (const [key, value] of Object.entries(props)) {
		attrs += `${key}="${value}" `
	}

	return html`
		<${unsafeStatic(tag)} ${unsafeStatic(attrs)}>
			${children ?? nothing}
		</${unsafeStatic(tag)}>
	`
}
```

The usage would be as follows:

```typescript
@customElement('hello-world')
class HelloWorld extends LitElement {
	@property()
	name: string

	@property({ type: Number })
	age: number

	static html(name: string, age: number, children: TemplateResult<1>) {
		return el('hello-world', { name, age }, children)
	}

	render() {
		return html`
			<div style="border: 1px solid black; padding: 5px">
				<p>Hello world component</p>
				<div>
					children:
					<slot></slot>
				</div>
				<p>data:</p>
				<p>name: ${this.name}</p>
				<p>age: ${this.age}</p>
			</div>
		`
	}
}

// In some other component:

	render() {
		return html`
			<div>
				${HelloWorld.html('aaa', 1234, html`
					<p> child elements </p>
				`)}
			</div>
		`
	}

```

I would like to know your opinion about this. This allows me to have a better interface about how do I create the element, and I get a better IDE experience.
However, I fear that the use of unsafeStatic won't be supported in future lit versions, or that I'm doing something that is really unperformant.

What are your thoughts on this?

**Comment by justinfagnani** [comment]
This would be really slow. Every time you change an attribute the entire DOM will be removed and replaced.

**Reply by alarbada** [reply]
What if I do this:


```typescript
@customElement('hello-world')
class HelloWorld extends LitElement {
	@property()
	name: string

	@property({ type: Number })
	age: number

	constructor(name: string, age: number) {
		super()
		this.name = name
		this.age = age
	}

	render() {
		return html`
			<div>
				name: ${this.name}
				age: ${this.age}
			</div>
		`
	}
}

// in consumer code

render() {
	return html`
		<div>
			${new HelloWorld('aaa', 1234)}
		</div>
	`
}

```

Isn't there a way to recreate jsx in lit, basically?

**Reply by augustjk** [reply]
Encouraging having to instantiate a custom element class this way to pass arguments into the constructor doesn't feel right and wouldn't cover any inherited properties.

Trying to use JSX for templating also hits its limitation that there's no declarative way to differentiate between properties and attributes, nor to add event listeners.
